src/d3a/d3a_core/sim_results/device_statistics.py

```diff
@@ -27,155 +20,188 @@ def _calc_min_max_from_sim_dict(subdict: Dict, key: str):
+            cls.calculate_stats_for_device(area, subdict, core_stats, current_market_time_slot)
 
     @classmethod
-    def calculate_stats_for_device(cls, area, market, subdict):
+    def calculate_stats_for_device(cls, area, subdict, core_stats, current_market_time_slot):
         key_name = "trade_energy_kWh"
+        if core_stats[area['uuid']] == {}:
+            return
+        area_core_trades = core_stats[area['uuid']].get('trades', [])
+
         traded_energy = 0
-        for t in market.trades:
-            if t.seller == area.name:
-                traded_energy -= t.offer.energy
-            if t.buyer == area.name:
-                traded_energy += t.offer.energy
+        for t in area_core_trades:
+            if t['seller'] == area['name']:
+                traded_energy -= t['energy']
+            if t['buyer'] == area['name']:
+                traded_energy += t['energy']
 
-        create_or_update_subdict(subdict, key_name, {market.time_slot: traded_energy})
+        create_or_update_subdict(
+            subdict, key_name,
+            {current_market_time_slot: traded_energy})
         cls._calc_min_max_from_sim_dict(subdict, key_name)
 
     @classmethod
-    def calculate_stats_for_infinite_bus(cls, area, market, subdict):
+    def calculate_stats_for_infinite_bus(cls, area, subdict, core_stats, current_market_time_slot):
         sold_key_name = "sold_trade_energy_kWh"
         bought_key_name = "bought_trade_energy_kWh"
         sold_traded_energy = 0
         bought_traded_energy = 0
-
-        for t in market.trades:
-            if t.seller == area.name:
-                sold_traded_energy += t.offer.energy
-            if t.buyer == area.name:
-                bought_traded_energy += t.offer.energy
-
-        create_or_update_subdict(subdict, sold_key_name, {market.time_slot: sold_traded_energy})
+        if core_stats[area['uuid']] == {}:
+            return
+        area_core_trades = core_stats[area['uuid']].get('trades', [])
+
+        for t in area_core_trades:
+            if t['seller'] == area['name']:
+                sold_traded_energy += t['energy']
+            if t['buyer'] == area['name']:
+                bought_traded_energy += t['energy']
+        create_or_update_subdict(
+            subdict, sold_key_name,
+            {current_market_time_slot: sold_traded_energy})
         cls._calc_min_max_from_sim_dict(subdict, sold_key_name)
-        create_or_update_subdict(subdict, bought_key_name,
-                                 {market.time_slot: bought_traded_energy})
+        create_or_update_subdict(
+            subdict, bought_key_name,
+            {current_market_time_slot: bought_traded_energy})
         cls._calc_min_max_from_sim_dict(subdict, bought_key_name)
 
     @classmethod
-    def _pv_production_stats(cls, area: Area, subdict: Dict):
+    def _pv_production_stats(cls, area: Dict, subdict: Dict, core_stats={},
+                             current_market_time_slot=None):
```

Comment:
Very minor, but to keep the variable names descriptive but short, I would say that `current_market_slot` should be sufficient.

Reply:
Ok sir!